src/v/cloud_topics/level_one/compaction/sink.cc

```diff
@@ -162,20 +145,119 @@ ss::future<> compaction_sink::maybe_roll() {
+          builder->finish());
+        co_await builder->close();
+        if (object_info_fut.failed()) {
+            auto e = object_info_fut.get_exception();
+            vlogl(
+              compaction_log,
+              ssx::is_shutdown_exception(e) ? ss::log_level::warn
+                                            : ss::log_level::error,
+              "Exception creating object_info: {}. Exiting compaction early.",
+              e);
+            co_await active_staging_file->remove();
+            std::rethrow_exception(e);
+        }
+
+        auto object_info = object_info_fut.get();
+        auto ntp_md = [this](const object_builder::object_info& info) {
+            auto [first, last] = info.index.partitions.equal_range(_tp);
+            vassert(
+              std::distance(first, last) == 1,
+              "Expected one partition range in builder.");
+            return metastore::object_metadata::ntp_metadata{
+              .tidp = _tp,
+              .base_offset = _object_base_offset,
+              .last_offset = _last_processed_offset,
+              .max_timestamp = first->second.max_timestamp,
+              .pos = first->second.file_position,
+              .size = first->second.length};
+        }(object_info);
+
+        auto file_and_info = file_and_md_info{
+          .staging_file = std::move(active_staging_file),
+          .info = std::move(object_info),
+          .ntp_md = std::move(ntp_md)};
+
+        // TODO: push update to committer.
+        std::ignore = std::move(file_and_info);
+
+        _object_base_offset = kafka::next_offset(_last_processed_offset);
+    }
+
+    // 2. Start new `_active_staging_file` and `_builder`.
+    if (initialize_new_builder) {
+        co_await initialize_builder();
+    }
+}
 
-    co_return;
+ss::future<> compaction_sink::maybe_roll() {
+    if (needs_roll()) {
+        co_await roll(true);
+    }
 }
 
 ss::future<ss::stop_iteration>
 compaction_sink::operator()(model::record_batch b, model::compression c) {
+    // Set last processed offset _before_ attempting to roll the object.
+    auto prev_offset = std::max(
+      kafka::prev_offset(model::offset_cast(b.base_offset())),
+      kafka::offset{0});
+    set_last_processed_offset(prev_offset);
+
     co_await maybe_roll();
+
     if (c != model::compression::none) {
         b = co_await model::compress_batch(c, std::move(b));
     }
+
     co_await _builder->add_batch(std::move(b));
+
     co_return ss::stop_iteration::no;
 }
 
-ss::future<> compaction_sink::finalize() { co_await commit_update_and_roll(); }
+ss::future<> compaction_sink::process_next_extent_offset_bounds(
+  kafka::offset next_extent_base, kafka::offset next_extent_last) {
+    bool is_first_extent = _object_base_offset == kafka::offset{};
+    if (is_first_extent) {
+        _object_base_offset = next_extent_base;
+    } else {
+        _processed_extents.insert(_extent_base_offset, _extent_last_offset);
```

Comment:
These are input extent offsets

Reply:
I think we are fundamentally keeping track of two things:
- the currently pending object
- the total set of input extents that we have seen